### PR TITLE
Add app health policies param for config upgrade

### PR DESCRIPTION
**Pull request by jkochhar**
Add AppHealthPolicyMap parameter to Swagger and CLI for ClusterConfigurationUpgradeDescription

Changes include:
Update the definition of configuration_upgrade to be able to specify application health policies during config upgrade
Update the CustomSDK from the latest in Swagger

Verifed: By building and running verify script

- [X] Build and test CI passes
- [x] History and readme updated to reflect changes
- [ ] Package version updated according to semantic versioning rules
- [X] Tests modified or added, when applicable
- [X] Updated code owners file, when applicable
- [X] Read the [PR checklist](https://github.com/Azure/service-fabric-cli/wiki/Checklist)

### CODE REVIEW

File: src/sfctl/custom_cluster_upgrade.py [hunk 1]
```diff
@@ -142,7 +144,8 @@ def sa_configuration_upgrade( #pylint: disable=missing-docstring,invalid-name,to
         max_percent_unhealthy_applications=unhealthy_applications,
         max_percent_unhealthy_nodes=unhealthy_nodes,
         max_percent_delta_unhealthy_nodes=delta_unhealthy_nodes,
-        max_percent_upgrade_domain_delta_unhealthy_nodes=upgrade_domain_delta_unhealthy_nodes) #pylint: disable=line-too-long
+        max_percent_upgrade_domain_delta_unhealthy_nodes=upgrade_domain_delta_unhealthy_nodes, #pylint: disable=line-too-long
```
Comment (samedder) [inline src/sfctl/custom_cluster_upgrade.py]: Don't disable line too long, use different variable names.

File: src/sfctl/custom_cluster_upgrade.py [hunk 2]
```diff
@@ -125,14 +125,16 @@ def upgrade( #pylint: disable=too-many-locals,missing-docstring,invalid-name,too
...
     client.start_cluster_upgrade(upgrade_desc, timeout=timeout)
 
-def sa_configuration_upgrade( #pylint: disable=missing-docstring,invalid-name,too-many-arguments
+def sa_configuration_upgrade( #pylint: disable=missing-docstring,invalid-name,too-many-arguments,too-many-locals
         client, cluster_config, health_check_retry='PT0H0M0S',
         health_check_wait='PT0H0M0S', health_check_stable='PT0H0M0S',
         upgrade_domain_timeout='PT0H0M0S', upgrade_timeout='PT0H0M0S',
         unhealthy_applications=0, unhealthy_nodes=0, delta_unhealthy_nodes=0,
-        upgrade_domain_delta_unhealthy_nodes=0, timeout=60):
+        upgrade_domain_delta_unhealthy_nodes=0, application_health_policies=None, timeout=60):
```
Comment (samedder) [inline src/sfctl/custom_cluster_upgrade.py]: Is this a JSON encoded list? If so you need to also add an argument parser.

File: src/sfctl/custom_cluster_upgrade.py [hunk 3]
```diff
@@ -125,14 +125,16 @@ def upgrade( #pylint: disable=too-many-locals,missing-docstring,invalid-name,too
...
     from azure.servicefabric.models import ClusterConfigurationUpgradeDescription
 
+    app_health_policies = parse_app_health_policy(application_health_policies)
+
     upgrade_desc = ClusterConfigurationUpgradeDescription(
```
Comment (samedder) [inline src/sfctl/custom_cluster_upgrade.py]: I assume this depends on the 6.2 SDK, correct?

File: src/sfctl/custom_cluster_upgrade.py (hunk 2 above)
Comment (jkochhar) [reply src/sfctl/custom_cluster_upgrade.py]: There exists a parser for it already.. parse_app_health_policy if that's what you mean

File: src/sfctl/custom_cluster_upgrade.py (hunk 3 above)
Comment (jkochhar) [reply src/sfctl/custom_cluster_upgrade.py]: 6.2+ These changes are not for 6.2 release as the corresponding Swagger and client changes are not in 6.2

File: src/sfctl/custom_cluster_upgrade.py (hunk 1 above)
Comment (jkochhar) [reply src/sfctl/custom_cluster_upgrade.py]: I didn't.. this is preexisting

File: src/sfctl/tests/mock_server.py [hunk 4]
```diff
@@ -30,7 +30,7 @@ def do_PUSH(self):  # pylint: disable=C0103,missing-docstring
     def do_POST(self):  # pylint: disable=C0103,missing-docstring
         # Certain requests expect a very specific response.
         # For those, return other status codes
-        if self.path.startswith('/$/RollbackUpgrade?') or self.path.startswith('/ComposeDeployments/deploymentName/$/Delete'):  # pylint: disable=line-too-long
+        if self.path.startswith('/$/RollbackUpgrade?') or self.path.startswith('/ComposeDeployments/deploymentName/$/Delete') or self.path.startswith('/$/StartClusterConfigurationUpgrade'):  # pylint: disable=line-too-long
```
Comment (samedder) [inline src/sfctl/tests/mock_server.py]: Use multi line if with `(` `)` to avoid using the `disable=line-too-long` expression

Comment (jkochhar) [reply src/sfctl/tests/mock_server.py]: Was preexisting.. but have removed the disable using your suggestion.

File: src/sfctl/custom_cluster_upgrade.py (hunk 2 above)
Comment (samedder) [reply src/sfctl/custom_cluster_upgrade.py]: No so, if it's a new JSON encoded argument you need to add an argument parser. All arguments are treated as string unless there's a [line like this](https://github.com/Azure/service-fabric-cli/blob/b22420189d81d9bb4ca6040c5953058cd3efd04d/src/sfctl/params.py#L180) in the params file. Please be sure to add one.

Comment (jkochhar) [reply src/sfctl/custom_cluster_upgrade.py]: Oh ok.. I think I needed that while running tests etc so added it. See params.py

File: CODEOWNERS [hunk 5]
```diff
@@ -43,3 +43,7 @@ src/sfctl/custom_chaos_schedule.py @likevi
 
 # Container related
 src/sfctl/custom_container.py @leikong
+
+# Upgrade related
```
Comment (Christina-Kang) [inline CODEOWNERS]: nit: can you name this Cluster Upgrade related please? Thanks!

File: src/sfctl/helps/cluster_upgrade.py [hunk 6]
```diff
@@ -165,6 +165,13 @@
           short-summary: The maximum allowed percentage of upgrade domain delta
             health degradation during the upgrade. Allowed values are integer
             values from zero to 100.
+        - name: --application-health-policies
+          type: string
+          short-summary: JSON encoded dictionary of pairs of application type
+            name and maximum percentage unhealthy before raising error
+    examples:
+        - name: Start a cluster configuration update
+          text: sfctl sa-cluster config-upgrade --cluster-config <YOUR CLUSTER CONFIG>
```
Comment (Christina-Kang) [inline src/sfctl/helps/cluster_upgrade.py]: What format is the value of `cluster config`? Can you include a concrete example? Same for the new param `application-health-policies`. What format should the JSON be?

 Can you also improve the help text for cluster config? "The cluster configuration" is not all that helpful :) Thanks!!

File: src/sfctl/tests/request_generation_test.py [hunk 7]
```diff
@@ -788,3 +788,14 @@ def paths_generation_helper(self):  # pylint: disable=too-many-statements
             'GET',
             '/Tools/Chaos',
             ['api-version=6.2'])
+
+        # Cluster Upgrade Commands:
+        self.validate_command(  # config-upgrade
```
Comment (Christina-Kang) [inline src/sfctl/tests/request_generation_test.py]: Please also include the other variables of the command in this test. Thanks!

File: CODEOWNERS (hunk 5 above)
Comment (jkochhar) [reply CODEOWNERS]: Sure

File: src/sfctl/tests/request_generation_test.py (hunk 7 above)
Comment (Christina-Kang) [reply src/sfctl/tests/request_generation_test.py]: Nope, this is the only validation for this particular command. That this was not there before was a gap in testing

File: src/sfctl/helps/cluster_upgrade.py (hunk 6 above)
Comment (Christina-Kang) [reply src/sfctl/helps/cluster_upgrade.py]: `This is exactly what I was saying.. the config is huge and it doesn't make sense to put the full json blob here.` Can you find an example elsewhere online then? If so, maybe we should include a link to it, or a mention of it, so users have an easier time finding an example somewhere. I think that users might know what it is and have a rough idea, but having the exact syntax down is another thing

**Comment by Christina-Kang** [comment]
@jkochhar Can you fix the merge conflict here as well? Thanks!

File: src/sfctl/custom_cluster_upgrade.py (hunk 1 above)
Comment (samedder) [reply src/sfctl/custom_cluster_upgrade.py]: Oh gotchya, okay

File: src/sfctl/tests/request_generation_test.py (hunk 7 above)
Comment (jkochhar) [reply src/sfctl/tests/request_generation_test.py]: but I don't see a point in adding other variables with default values..

Comment (jkochhar) [reply src/sfctl/tests/request_generation_test.py]: Done

**Comment by samedder** [comment]
@Christina-Kang can you look at the chaos commands? It seems to have a parsing error now for the timeout option:

> AssertionError: ERROR while running command "chaos schedule set --version 0 --start-date-utc 2016-01-01T00:00:00.000Z --expiry-date-utc 2038-01-01T00:00:00.000Z --chaos-parameters-dictionary [{\"Key\":\"adhoc\",\"Value\":{\"MaxConcurrentFaults\":3,\"EnableMoveReplicaFaults\":true,\"ChaosTargetFilter\":{\"NodeTypeInclusionList\":[\"N0010Ref\",\"N0020Ref\",\"N0030Ref\",\"N0040Ref\",\"N0050Ref\"]},\"MaxClusterStabilizationTimeoutInSeconds\":60,\"WaitTimeBetweenIterationsInSeconds\":15,\"WaitTimeBetweenFaultsInSeconds\":30,\"TimeToRunInSeconds\":\"600\",\"Context\":{\"Map\":{\"test\":\"value\"}},\"ClusterHealthPolicy\":{\"MaxPercentUnhealthyNodes\":0,\"ConsiderWarningAsError\":true,\"MaxPercentUnhealthyApplications\":0}}}] --jobs [{\"ChaosParameters\":\"adhoc\",\"Days\":{\"Sunday\":true,\"Monday\":true,\"Tuesday\":true,\"Wednesday\":true,\"Thursday\":true,\"Friday\":true,\"Saturday\":true},\"Times\":[{\"StartTime\":{\"Hour\":0,\"Minute\":0},\"EndTime\":{\"Hour\":23,\"Minute\":59}}]}]". Error: "The command failed. Exit code: 1"

**Comment by jkochhar** [comment]
It looks like the test failure is related to a change in Swagger for Chaos API by Kevin. I started a thread regarding that. If I skip that test, rest was clean.

File: src/sfctl/tests/request_generation_test.py (hunk 7 above)
Comment (Christina-Kang) [reply src/sfctl/tests/request_generation_test.py]: Thank you! :)

**Pull request merged**